### PR TITLE
ec2: add set-instance-profile action

### PR DESCRIPTION
**Pull request by jesselang**
It's useful to associate an instance profile to an existing instance.

### CODE REVIEW

**Comment by jesselang** [comment]
I'm finished verifying the docs and permissions. Should be ready to go. Thanks!

File: tests/test_ec2.py [hunk 1]
```diff
@@ -555,6 +555,23 @@ def test_ec2_snapshot_copy_tags(self):
...
+            'name': 'ec2-test-associate-instance-profile',
+            'resource': 'ec2',
+            'filters': [
+                {'tag:Name': 'MissingInstanceProfile'}],
+            'actions': [
+                {'type': 'associate-instance-profile',
+                 'name': 'ec2-default'}]},
+            session_factory=session_factory)
+        resources = policy.run()
+        self.assertEqual(len(resources), 1)
```
Comment (kapilt) [inline tests/test_ec2.py]: we've got a few tests that have some bad verifications...

to be a bit clearer about the test assertion, can you add a filter to the test policy that matches that the instance doesn't have an instance profile (or different) and then use an ec2 client describe_instances to verify that the role is present on the instance after the policy execution.

File: c7n/resources/ec2.py [hunk 2]
```diff
@@ -918,6 +918,43 @@ def process(self, instances):
...
+            actions:
+              - type: associate-instance-profile
+                name: default
+
+    https://docs.aws.amazon.com/cli/latest/reference/ec2/associate-iam-instance-profile.html
+    """
+
+    schema = type_schema(
+        'associate-instance-profile',
+        **{'name': {'type': 'string'}})
```
Comment (kapilt) [inline c7n/resources/ec2.py]: wondering if we shouldn't also have this have the ability to yank a profile if None or empty string is provided. thoughts?

**Comment by jesselang** [comment]
Good plan. On it.

**Comment by jesselang** [comment]
@kapilt I've made the requested changes. I would appreciate another look over when you get a chance. Thanks much!

**Comment by jesselang** [comment]
@kapilt the travis error being reported isn't clear to me. Seems it's missing a `config.json`. Is this environmental? `make test` passes when I run it in my sandbox.

File: c7n/resources/ec2.py [hunk 3]
```diff
@@ -918,6 +918,67 @@ def process(self, instances):
...
+            self.manager.session_factory).client('ec2')
+        profile_name = self.data.get('name', '')
+
+        for i in instances:
+            if profile_name:
+                client.associate_iam_instance_profile(
+                    IamInstanceProfile={'Name': self.data.get('name', '')},
+                    InstanceId=i['InstanceId'])
+            else:
+                response = client.describe_iam_instance_profile_associations(
```
Comment (kapilt) [inline c7n/resources/ec2.py]: instance profile is already available as an attribute of the instance per describe output, as IamInstanceProfile

**Comment by kapilt** [comment]
thanks, looks like an unrelated issue, i did a quick through travis history couldn't find a previous occurrence in the last few days, a rebuild resolved.

File: c7n/resources/ec2.py (hunk 3 above)
Comment (jesselang) [inline c7n/resources/ec2.py]: Not sure I understand your comment. The instance profile is available as an instance attribute, however what we're after here is the association ID that can be passed to `disassociate_iam_instance_profile`. Could you clarify? Thanks!

Comment (kapilt) [inline c7n/resources/ec2.py]: ic. that makes sense re associate id

**Comment by jesselang** [comment]
Hey @kapilt. Hope you had a delightful long weekend. Is there anything outstanding here that needs to be done before merging? Do let me know. Thanks!

**Comment by jesselang** [comment]
Thanks @kapilt !

**Pull request merged**